cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/security/trustedproxy/TrustedProxyLoginService.java

```diff
@@ -0,0 +1,151 @@
+ * based on the {@link AuthorizationService}.
+ */
+public class TrustedProxyLoginService extends ContainerLifeCycle implements LoginService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TrustedProxyLoginService.class);
+  private final AuthorizationService _trustedUserAuthorizer;
+  // use encapsulation instead of inheritance as it's easier to test
+  private final ConfigurableSpnegoLoginService _delegateSpnegoLoginService;
+
+  /**
+   * Creates a new instance based on the kerberos realm, the list of trusted proxies, their allowed IP pattern and the
+   * {@link AuthorizationService} that stores the end users.
+   * @param realm
+   * @param userAuthorizer
+   * @param trustedProxies
+   * @param trustedProxyIpPattern
+   */
+  public TrustedProxyLoginService(String realm, AuthorizationService userAuthorizer, List<String> trustedProxies, String trustedProxyIpPattern) {
+    _delegateSpnegoLoginService = new SpnegoLoginServiceWithAuthServiceLifecycle(
+        realm, new TrustedProxyAuthorizationService(trustedProxies, trustedProxyIpPattern));
+    _trustedUserAuthorizer = userAuthorizer;
+  }
+
+  // visible for testing
+  TrustedProxyLoginService(ConfigurableSpnegoLoginService delegateSpnegoLoginService, AuthorizationService trustedProxyAuthorizer) {
+    _delegateSpnegoLoginService = delegateSpnegoLoginService;
+    _trustedUserAuthorizer = trustedProxyAuthorizer;
+  }
+
+  // ------- ConfigurableSpnegoLoginService methods -------
+
+  public void setServiceName(String serviceName) {
+    _delegateSpnegoLoginService.setServiceName(serviceName);
+  }
+
+  public void setHostName(String hostName) {
+    _delegateSpnegoLoginService.setHostName(hostName);
+  }
+
+  public void setKeyTabPath(Path path) {
+    _delegateSpnegoLoginService.setKeyTabPath(path);
+  }
+
+  // ------- LoginService methods -------
+
+  @Override
+  public String getName() {
+    return _delegateSpnegoLoginService.getName();
+  }
+
+  @Override
+  public UserIdentity login(String username, Object credentials, ServletRequest request) {
+    if (!(request instanceof HttpServletRequest)) {
+      return null;
+    }
+    SpnegoUserIdentity serviceIdentity = (SpnegoUserIdentity) _delegateSpnegoLoginService.login(username, credentials, request);
+    SpnegoUserPrincipal servicePrincipal = (SpnegoUserPrincipal) serviceIdentity.getUserPrincipal();
+    String doAsUser = request.getParameter(DO_AS);
+    LOG.info("Authorizing proxy user {} from {} service", doAsUser, servicePrincipal.getName());
+    UserIdentity doAsIdentity = null;
+    if (doAsUser != null && !doAsUser.isEmpty()) {
+      doAsIdentity = _trustedUserAuthorizer.getUserIdentity((HttpServletRequest) request, doAsUser);
+    }
+
+    Principal principal = new TrustedProxyPrincipal(doAsUser, servicePrincipal);
+    Subject subject = new Subject(true, Collections.singleton(principal), Collections.emptySet(), Collections.emptySet());
```

Comment:
Nit: Public/private credentials of `subject` seem to use some default values, should we move them to static variables?